### PR TITLE
ModernBERT bug fixes

### PR DESCRIPTION
**Pull request by warner-benjamin**
This PR fixes a few issues with the ModernBERT implementation and some typos in the docs.

First, on the flash attention 2 path, ModernBERT was incorrectly always returning repadded logits without gradients (see #35386). This PR will return padded logits with gradients if no labels are passed to `ModernBertForMaskedLM` or if `repad_logits_with_grad` is set to true. We don't keep the gradient when repadding after using an internal loss function by default to save memory.

Second, this PR fixes torch.compile from being automatically set on when on CPU (#35388).

Third, I added some details to the model doc strings.

Fourth, documentation is updated to capitalize BERT in ModernBERT following the other BERT models.

cc @ArthurZucker @tomaarsen

### CODE REVIEW

File: docs/source/en/_toctree.yml [hunk 1]
```diff
@@ -503,7 +503,7 @@
       - local: model_doc/mobilebert
         title: MobileBERT
       - local: model_doc/modernbert
-        title: ModernBert
+        title: ModernBERT
```
Comment (tomaarsen) [inline docs/source/en/_toctree.yml:506]: Good call - I got carried away with the Python class naming

File: src/transformers/models/modernbert/modeling_modernbert.py [hunk 2]
```diff
@@ -667,6 +670,13 @@ def _maybe_set_compile(self):
                 )
             self.config.reference_compile = False
 
+        if self.device.type == "cpu":
+            logger.warning_once(
```
Comment (tomaarsen) [inline src/transformers/models/modernbert/modeling_modernbert.py:674]: I think we should wrap this in `if self.config.reference_compile:` just like above so users don't get warned when they never manually specified anything related to `reference_compile`.

Unless you'd rather always give this warning so people aren't confused why there's a small precision-esque difference between GPU setups?

File: src/transformers/models/modernbert/modular_modernbert.py [hunk 3]
```diff
@@ -892,6 +900,13 @@ def _maybe_set_compile(self):
                 )
             self.config.reference_compile = False
 
+        if self.device.type == "cpu":
+            logger.warning_once(
```
Comment (tomaarsen) [inline src/transformers/models/modernbert/modular_modernbert.py:904]: I think we should wrap this in `if self.config.reference_compile:` just like above so users don't get warned when they never manually specified anything related to `reference_compile`.

Unless you'd rather always give this warning so people aren't confused why there's a small precision-esque difference between GPU setups?

File: src/transformers/models/modernbert/modular_modernbert.py [hunk 4]
```diff
@@ -852,12 +858,14 @@ def _autoset_attn_implementation(
...
         # Otherwise we fall back to the default SDPA -> Eager from the super() method.
+        # ModernBert's FA2 implementation correctly handles non-fp16/bf16 dtypes, we don't
+        # need the FA2 warning for non-fp16/bf16 dtypes so we set fp16 for the FA2 check.
         if config._attn_implementation_internal is None:
             config._attn_implementation_internal = "flash_attention_2"
             try:
                 return cls._check_and_enable_flash_attn_2(
                     config,
-                    torch_dtype=torch_dtype,
+                    torch_dtype=torch.float16,
```
Comment (tomaarsen) [inline src/transformers/models/modernbert/modular_modernbert.py:868]: I think this is a cleaner solution to avoid the unnecessary FP32 warning than I figured was possible, nice.

File: src/transformers/models/modernbert/modular_modernbert.py (hunk 3 above)
Comment (warner-benjamin) [reply src/transformers/models/modernbert/modular_modernbert.py]: That was my intention, I forgot the if statement. Fixed in ab11657.

Comment (tomaarsen) [reply src/transformers/models/modernbert/modular_modernbert.py]: Nice! Resolved

**Review by tomaarsen — APPROVED** [review]

File: src/transformers/models/modernbert/modular_modernbert.py [hunk 5]
```diff
@@ -141,6 +142,9 @@ class ModernBertConfig(PretrainedConfig):
             the model will be compiled if 1) `triton` is installed, 2) the model is not on MPS, 3) the model is not
             shared between devices, and 4) the model is not resized after initialization. If `True`, then the model may
             be faster in some scenarios.
+        repad_logits_with_grad (`bool`, *optional*, defaults to `False`):
+            When True, ModernBertForMaskedLM keep track of the logits' gradient when repadding for output. This only
```
Comment (tomaarsen) [inline src/transformers/models/modernbert/modular_modernbert.py:146]:
```suggestion
            When True, ModernBertForMaskedLM keeps track of the logits' gradient when repadding for output. This only
```

**Review by ArthurZucker — COMMENTED** [review]
I'll let @tomaarsen update, I think we can just check if training and labels iwthout having a new arg

File: src/transformers/models/modernbert/modular_modernbert.py [hunk 6]
```diff
@@ -1246,8 +1257,9 @@ def forward(
             loss = self.loss_function(logits, labels, vocab_size=self.config.vocab_size)
 
         if self.config._attn_implementation == "flash_attention_2":
-            with torch.no_grad():
+            with nullcontext() if self.training or labels is None else torch.no_grad():
```
Comment (tomaarsen) [inline src/transformers/models/modernbert/modular_modernbert.py:1260]: @warner-benjamin I just updated the `config.repad_logits_with_grad` with `self.training`, but I worry that this conditional doesn't make sense. Could you look it over? 
I don't see when we might need grads when not training, so perhaps this can just be:
```suggestion
            with nullcontext() if self.training else torch.no_grad():
```

**Review by ArthurZucker — APPROVED** [review]
LGTM, thanks both!

File: src/transformers/models/modernbert/modular_modernbert.py (hunk 6 above)
Comment (tomaarsen) [reply src/transformers/models/modernbert/modular_modernbert.py]: I've reversed this to avoid unnecessarily increasing memory when training in some situations.

**Review by ArthurZucker — APPROVED** [review]
